src/algorithms/calorimetry/CalorimeterHitsMerger.cc

```diff
@@ -41,24 +43,57 @@ void CalorimeterHitsMerger::init() {
         return;
     }
 
+    // initialize descriptor + decoders
     try {
-        auto id_desc = m_detector->readout(m_cfg.readout).idSpec();
-        id_mask = 0;
-        std::vector<std::pair<std::string, int>> ref_fields;
-        for (size_t i = 0; i < m_cfg.fields.size(); ++i) {
-            id_mask |= id_desc.field(m_cfg.fields[i])->mask();
-            // use the provided id number to find ref cell, or use 0
-            int ref = i < m_cfg.refs.size() ? m_cfg.refs[i] : 0;
-            ref_fields.emplace_back(m_cfg.fields[i], ref);
-        }
-        ref_mask = id_desc.encode(ref_fields);
+      id_desc = m_detector->readout(m_cfg.readout).idSpec();
+      id_decoder = id_desc.decoder();
+      for (const auto& field : m_cfg.fields) {
+        const short index = id_decoder->index(field);
+      }
     } catch (...) {
-        auto mess = fmt::format("Failed to load ID decoder for {}", m_cfg.readout);
-        warning(mess);
+      auto mess = fmt::format("Failed to load ID decoder for {}", m_cfg.readout);
+      warning(mess);
 //        throw std::runtime_error(mess);
     }
-    id_mask = ~id_mask;
-    debug("ID mask in {:s}: {:#064b}", m_cfg.readout, id_mask);
+
+    // if no field-by-field mappings provided, initialize relevant bitmasks
+    // otherwise intialize relevant functionals
+    if (m_cfg.mappings.empty()) {
+      id_mask = 0;
+      std::vector<RefField> ref_fields;
+      for (size_t i = 0; i < m_cfg.fields.size(); ++i) {
+          id_mask |= id_desc.field(m_cfg.fields[i])->mask();
+          // use the provided id number to find ref cell, or use 0
+          int ref = i < m_cfg.refs.size() ? m_cfg.refs[i] : 0;
+          ref_fields.emplace_back(m_cfg.fields[i], ref);
+      }
+      ref_mask = id_desc.encode(ref_fields);
+      id_mask = ~id_mask;
+      debug("ID mask in {:s}: {:#064b}", m_cfg.readout, id_mask);
```

Comment:
Can we put mask operations as closures to the same ref_maps, so that the application of transofrmation is uniform?

Reply:
I don't quite follow... So by closure do you mean if a user provides both a bitmask and a mapping, they both get applied?

Reply:
Allow me to clarify. I meant that this code can be replaced with something like
```
for (size_t ix = 0; std::string &field : m_cfg.fields) {
  uint64_t ref = m_cfg.refs.at(ix);
  ref_maps[field] = [ref](const edm4eic::CalorimeterHit&) -> int {
    return ref;
  };
  ix++;
}
```
Then we can remove all the different logic for handling ref_mask and id_mask.

Reply:
I think this it's more like:
```
for (size_t ix = 0; std::string &field : m_cfg.fields) {
  uint64_t ref_mask = m_cfg.refs.at(ix);
  auto field_element = id_desc.field(m_cfg.fields[i]);
  ref_maps[field] = [ref_mask,field_element](const edm4eic::CalorimeterHit& hit) -> int {
    return (field_element->value(hit.cellID()) & ~ref_mask) | ref_mask;
  };
  ix++;
}
```

Reply:
I see! And agreed! That would help to simplify things... I'll work through the implications for the rest of the algorithm...